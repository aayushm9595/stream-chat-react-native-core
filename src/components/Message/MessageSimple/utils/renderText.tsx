import React from 'react';
// @ts-expect-error
import Markdown from '@stream-io/react-native-simple-markdown';
import anchorme from 'anchorme';
import truncate from 'lodash/truncate';

import type { MarkdownStyle } from '../../../../styles/themeConstants';
import type { Message } from '../../../MessageList/utils/insertDates';
import type {
  DefaultAttachmentType,
  DefaultChannelType,
  DefaultCommandType,
  DefaultEventType,
  DefaultMessageType,
  DefaultReactionType,
  DefaultUserType,
  UnknownType,
} from '../../../../types/types';

const defaultMarkdownStyles = {
  inlineCode: {
    backgroundColor: '#F3F3F3',
    borderColor: '#dddddd',
    color: 'red',
    fontSize: 13,
    padding: 3,
    paddingHorizontal: 5,
  },
  link: {
    color: 'blue',
    textDecorationLine: 'underline',
  },
  url: {
    color: 'blue',
    textDecorationLine: 'underline',
  },
};

export type RenderTextParams<
  At extends UnknownType = DefaultAttachmentType,
  Ch extends UnknownType = DefaultChannelType,
  Co extends string = DefaultCommandType,
  Ev extends UnknownType = DefaultEventType,
  Me extends UnknownType = DefaultMessageType,
  Re extends UnknownType = DefaultReactionType,
  Us extends UnknownType = DefaultUserType
> = {
  markdownRules: UnknownType;
  markdownStyles: MarkdownStyle;
  message: Message<At, Ch, Co, Ev, Me, Re, Us>;
};

export const renderText = <
  At extends UnknownType = DefaultAttachmentType,
  Ch extends UnknownType = DefaultChannelType,
  Co extends string = DefaultCommandType,
  Ev extends UnknownType = DefaultEventType,
  Me extends UnknownType = DefaultMessageType,
  Re extends UnknownType = DefaultReactionType,
  Us extends UnknownType = DefaultUserType
>(
  params: RenderTextParams<At, Ch, Co, Ev, Me, Re, Us>,
) => {
  const { markdownRules, markdownStyles, message } = params;

  // take the @ mentions and turn them into markdown?
  // translate links
  const { mentioned_users = [], text } = message;

  if (!text) return null;

  let newText = text.trim();
  const urls = anchorme(newText, {
    list: true,
  });

  for (const urlInfo of urls) {
    const displayLink = truncate(urlInfo.encoded.replace(/^(www\.)/, ''), {
      length: 200,
      omission: '...',
    });
    const markdown = `[${displayLink}](${urlInfo.protocol}${urlInfo.encoded})`;
    newText = newText.replace(urlInfo.raw, markdown);
  }

  if (mentioned_users.length) {
    for (let i = 0; i < mentioned_users.length; i++) {
      const username = mentioned_users[i].name || mentioned_users[i].id;
      const markdown = `**@${username}**`;
      const regEx = new RegExp(`@${username}`, 'g');
      newText = newText.replace(regEx, markdown);
    }
  }

  newText = newText.replace(/[<&"'>]/g, '\\$&');
  const styles = {
    ...defaultMarkdownStyles,
    ...markdownStyles,
  };

  return (
    <Markdown rules={markdownRules} styles={styles}>
      {newText}
    </Markdown>
  );
};
