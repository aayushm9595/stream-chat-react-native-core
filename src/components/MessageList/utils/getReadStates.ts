import {
  InsertDatesResponse,
  isDateSeparator,
  Message,
} from '../utils/insertDates';

import type { ImmutableDate } from 'seamless-immutable';
import type { UserResponse } from 'stream-chat';

import type { ChannelContextValue } from '../../../contexts/channelContext/ChannelContext';
import type {
  DefaultAttachmentType,
  DefaultChannelType,
  DefaultCommandType,
  DefaultEventType,
  DefaultMessageType,
  DefaultReactionType,
  DefaultUserType,
  UnknownType,
} from '../../../types/types';

export const getReadStates = <
  At extends UnknownType = DefaultAttachmentType,
  Ch extends UnknownType = DefaultChannelType,
  Co extends string = DefaultCommandType,
  Ev extends UnknownType = DefaultEventType,
  Me extends UnknownType = DefaultMessageType,
  Re extends UnknownType = DefaultReactionType,
  Us extends UnknownType = DefaultUserType
>(
  messages: InsertDatesResponse<At, Ch, Co, Ev, Me, Re, Us>,
  read?: ChannelContextValue<At, Ch, Co, Ev, Me, Re, Us>['read'],
) => {
  const readData = messages.reduce((acc, cur) => {
    if (!isDateSeparator(cur) && cur.id) {
      acc[cur.id] = [];
    }
    return acc;
  }, {} as { [key: string]: UserResponse<Us>[] });

  const filteredMessagesReversed = messages
    .filter((msg) => !isDateSeparator(msg) && msg.updated_at)
    .reverse() as Array<
    Message<At, Ch, Co, Ev, Me, Re, Us> & {
      updated_at: string | ImmutableDate;
    }
  >;

  if (read) {
    /**
     * Channel read state is stored by user
     */
    for (const readState of Object.values(read)) {
      /**
       * If no last read continue
       */
      if (!readState.last_read) {
        continue;
      }

      /**
       * Array is in reverse order so newest message is at 0,
       * we find the index of the first message that is older
       * than the last read and then set last read to that, or
       * if there are no newer messages, the first message is
       * last read message.
       */
      const userLastReadMsgId = filteredMessagesReversed.find(
        (msg) => msg.updated_at < readState.last_read,
      )?.id;

      /**
       * If there there is a last read message add the user
       * to the array of last reads for that message
       */
      if (userLastReadMsgId) {
        readData[userLastReadMsgId] = [
          ...readData[userLastReadMsgId],
          readState.user as UserResponse<Us>,
        ];
      }
    }
  }

  return readData;
};
