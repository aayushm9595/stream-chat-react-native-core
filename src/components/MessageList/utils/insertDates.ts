import type { ChannelState } from 'stream-chat';

import type {
  MessagesContextValue,
  MessageWithDates,
} from '../../../contexts/messagesContext/MessagesContext';
import type { ThreadContextValue } from '../../../contexts/threadContext/ThreadContext';
import type {
  DefaultAttachmentType,
  DefaultChannelType,
  DefaultCommandType,
  DefaultEventType,
  DefaultMessageType,
  DefaultReactionType,
  DefaultUserType,
  UnknownType,
} from '../../../types/types';

export type DateSeparator<
  At extends UnknownType = DefaultAttachmentType,
  Ch extends UnknownType = DefaultChannelType,
  Co extends string = DefaultCommandType,
  Ev extends UnknownType = DefaultEventType,
  Me extends UnknownType = DefaultMessageType,
  Re extends UnknownType = DefaultReactionType,
  Us extends UnknownType = DefaultUserType
> = {
  date: Message<At, Ch, Co, Ev, Me, Re, Us>['created_at'] | string;
  type: 'message.date';
};

export type MessageOrDate<
  At extends UnknownType = DefaultAttachmentType,
  Ch extends UnknownType = DefaultChannelType,
  Co extends string = DefaultCommandType,
  Ev extends UnknownType = DefaultEventType,
  Me extends UnknownType = DefaultMessageType,
  Re extends UnknownType = DefaultReactionType,
  Us extends UnknownType = DefaultUserType
> =
  | Message<At, Ch, Co, Ev, Me, Re, Us>
  | DateSeparator<At, Ch, Co, Ev, Me, Re, Us>;

export type InsertDatesResponse<
  At extends UnknownType = DefaultAttachmentType,
  Ch extends UnknownType = DefaultChannelType,
  Co extends string = DefaultCommandType,
  Ev extends UnknownType = DefaultEventType,
  Me extends UnknownType = DefaultMessageType,
  Re extends UnknownType = DefaultReactionType,
  Us extends UnknownType = DefaultUserType
> = MessageOrDate<At, Ch, Co, Ev, Me, Re, Us>[];

export const isDateSeparator = <
  At extends UnknownType = DefaultAttachmentType,
  Ch extends UnknownType = DefaultChannelType,
  Co extends string = DefaultCommandType,
  Ev extends UnknownType = DefaultEventType,
  Me extends UnknownType = DefaultMessageType,
  Re extends UnknownType = DefaultReactionType,
  Us extends UnknownType = DefaultUserType
>(
  message: MessageOrDate<At, Ch, Co, Ev, Me, Re, Us>,
): message is DateSeparator<At, Ch, Co, Ev, Me, Re, Us> =>
  message.type === 'message.date';

export type Message<
  At extends UnknownType = DefaultAttachmentType,
  Ch extends UnknownType = DefaultChannelType,
  Co extends string = DefaultCommandType,
  Ev extends UnknownType = DefaultEventType,
  Me extends UnknownType = DefaultMessageType,
  Re extends UnknownType = DefaultReactionType,
  Us extends UnknownType = DefaultUserType
> =
  | ReturnType<ChannelState<At, Ch, Co, Ev, Me, Re, Us>['messageToImmutable']>
  | MessageWithDates<At, Ch, Co, Me, Re, Us>;

export const insertDates = <
  At extends UnknownType = DefaultAttachmentType,
  Ch extends UnknownType = DefaultChannelType,
  Co extends string = DefaultCommandType,
  Ev extends UnknownType = DefaultEventType,
  Me extends UnknownType = DefaultMessageType,
  Re extends UnknownType = DefaultReactionType,
  Us extends UnknownType = DefaultUserType
>(
  messages:
    | MessagesContextValue<At, Ch, Co, Ev, Me, Re, Us>['messages']
    | ThreadContextValue<At, Ch, Co, Ev, Me, Re, Us>['threadMessages'],
) => {
  const newMessages: InsertDatesResponse<At, Ch, Co, Ev, Me, Re, Us> = [];
  if (messages.length === 0) {
    return newMessages;
  }

  for (const [i, message] of messages.entries()) {
    /**
     * If message read or deleted don't consider for date labels
     */
    if (message.type === 'message.read' || message.deleted_at) {
      newMessages.push(message);
      continue;
    }

    /**
     * Get the date of the current message and create
     * variable for previous date (day)
     */
    const messageDate = message.created_at.getDay();

    /**
     * If this is not the last entry in the messages array
     * set the previous message date (day) to the date of the next
     * message in the array
     */
    const prevMessageDate =
      i < messages.length - 1
        ? messages[i + 1].created_at.getDay()
        : messageDate;

    /**
     * Before the first message insert a date object
     */
    if (i === 0) {
      newMessages.push(
        {
          date: message.created_at,
          type: 'message.date',
        } as DateSeparator<At, Ch, Co, Ev, Me, Re, Us>,
        message,
      );

      /**
       * If the date (day) has changed between two messages
       * insert a date object
       */
    } else if (messageDate !== prevMessageDate) {
      newMessages.push(message, {
        date: messages[i + 1].created_at,
        type: 'message.date',
      } as DateSeparator<At, Ch, Co, Ev, Me, Re, Us>);

      /**
       * Otherwise just add the message
       */
    } else {
      newMessages.push(message);
    }
  }

  return newMessages;
};
