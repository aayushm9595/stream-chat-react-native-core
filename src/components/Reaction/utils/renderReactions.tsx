import React from 'react';
import { Text } from 'react-native';

import type { ReactionResponse } from 'stream-chat';

import type { LatestReactions, Reaction } from '../ReactionList';
import type {
  DefaultAttachmentType,
  DefaultChannelType,
  DefaultCommandType,
  DefaultMessageType,
  DefaultReactionType,
  DefaultUserType,
  UnknownType,
} from '../../../types/types';

export const renderReactions = <
  At extends UnknownType = DefaultAttachmentType,
  Ch extends UnknownType = DefaultChannelType,
  Co extends string = DefaultCommandType,
  Me extends UnknownType = DefaultMessageType,
  Re extends UnknownType = DefaultReactionType,
  Us extends UnknownType = DefaultUserType
>(
  reactions: LatestReactions<At, Ch, Co, Me, Re, Us>,
  supportedReactions: Reaction[],
) => {
  const reactionsByType: {
    [key: string]: ReactionResponse<Re, Us>[];
  } = {};
  reactions?.map((reaction) => {
    if (!reactionsByType[reaction.type]) {
      return (reactionsByType[reaction.type] = [reaction]);
    }
    return (reactionsByType[reaction.type] = [
      ...reactionsByType[reaction.type],
      reaction,
    ]);
  });

  const emojiDataByType: { [key: string]: Reaction } = {};
  const reactionTypes = supportedReactions.map((supportedReaction) => {
    emojiDataByType[supportedReaction.id] = supportedReaction;
    return supportedReaction.id;
  });

  return Object.keys(reactionsByType).map((type) =>
    reactionTypes.indexOf(type) > -1 ? (
      <Text key={type} testID={type}>
        {emojiDataByType[type].icon}
      </Text>
    ) : null,
  );
};
