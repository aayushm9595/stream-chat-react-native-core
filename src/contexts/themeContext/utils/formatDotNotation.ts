import lodashGet from 'lodash/get';
import isPlainObject from 'lodash/isPlainObject';
import merge from 'lodash/merge';
import lodashSet from 'lodash/set';

import type { ThemeType } from './replaceCssShorthand';

import type { Theme } from '../../../styles/themeConstants';

export type FormatDotNotationParams = {
  formattedStyle: string | number | boolean | ThemeType;
  modifiedTheme: Theme;
};

export const formatDotNotation = ({
  formattedStyle,
  modifiedTheme,
}: FormatDotNotationParams) => {
  if (isPlainObject(formattedStyle)) {
    const themeDiff = {};
    for (const k in formattedStyle as ThemeType) {
      if (lodashGet(modifiedTheme, k)) {
        merge(themeDiff, lodashSet({}, k, (formattedStyle as ThemeType)[k]));
      } else {
        throw Error(`Unknown theme key ${k}`);
      }
    }
    return themeDiff;
  }
  return modifiedTheme;
};
