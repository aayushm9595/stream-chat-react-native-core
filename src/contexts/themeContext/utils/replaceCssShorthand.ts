import lodashGet from 'lodash/get';
import isPlainObject from 'lodash/isPlainObject';
import mapValues from 'lodash/mapValues';

import { defaultTheme } from '../../../styles/themeConstants';

export type ThemeType = {
  [key: string]: string | number | boolean | ThemeType;
};

// replaces
// { 'avatar.fallback': 'background-color: red;' }
// with
// { 'avatar.fallback': { css: 'background-color: red;' } }
export const replaceCssShorthand = (
  style: ThemeType | string | number | boolean,
): ThemeType | string | number | boolean => {
  const path: string[] = [];

  const replaceCssShorthandInternal = (
    styleInternal: ThemeType | string | number | boolean,
  ): ThemeType | string | number | boolean => {
    if (isPlainObject(styleInternal)) {
      const mapped = mapValues(styleInternal as ThemeType, (styleItem, key) => {
        path.push(key);
        return replaceCssShorthandInternal(styleItem);
      });
      path.pop();
      return mapped;
    }

    if (isPlainObject(lodashGet(defaultTheme, path.join('.')))) {
      path.pop();
      return { css: styleInternal };
    }
    path.pop();
    return styleInternal;
  };

  return replaceCssShorthandInternal(style);
};
