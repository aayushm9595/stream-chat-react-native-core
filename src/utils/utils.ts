import debounce from 'lodash/debounce';

import type {
  Channel,
  ChannelMemberAPIResponse,
  ChannelMemberResponse,
  UserResponse,
} from 'stream-chat';
import type { DebouncedFunc } from 'lodash';
import type { CommandsItemProps } from '../components/AutoCompleteInput/CommandsItem';
import type { MentionsItemProps } from '../components/AutoCompleteInput/MentionsItem';
import type {
  SuggestionCommand,
  SuggestionUser,
} from '../contexts/suggestionsContext/SuggestionsContext';
import type { TranslationContextValue } from '../contexts/translationContext/TranslationContext';
import type {
  DefaultAttachmentType,
  DefaultChannelType,
  DefaultCommandType,
  DefaultEventType,
  DefaultMessageType,
  DefaultReactionType,
  DefaultUserType,
  UnknownType,
} from '../types/types';

export const emojiData = [
  {
    icon: 'üëç',
    id: 'like',
  },
  {
    icon: '‚ù§Ô∏èÔ∏è',
    id: 'love',
  },
  {
    icon: 'üòÇ',
    id: 'haha',
  },
  {
    icon: 'üòÆ',
    id: 'wow',
  },
  {
    icon: 'üòî',
    id: 'sad',
  },
  {
    icon: 'üò†',
    id: 'angry',
  },
];

export const FileState = Object.freeze({
  // finished and uploaded state are the same thing. First is set on frontend,
  // while later is set on backend side
  // TODO: Unify both of them
  FINISHED: 'finished',
  NO_FILE: 'no_file',
  UPLOAD_FAILED: 'upload_failed',
  UPLOADED: 'uploaded',
  UPLOADING: 'uploading',
});

export const ProgressIndicatorTypes: {
  IN_PROGRESS: 'in_progress';
  RETRY: 'retry';
} = Object.freeze({
  IN_PROGRESS: 'in_progress',
  RETRY: 'retry',
});

const isUserResponse = <Us extends DefaultUserType = DefaultUserType>(
  user: SuggestionUser<Us> | undefined,
): user is SuggestionUser<Us> => (user as SuggestionUser<Us>) !== undefined;

const getCommands = <
  At extends UnknownType = DefaultAttachmentType,
  Ch extends UnknownType = DefaultChannelType,
  Co extends string = DefaultCommandType,
  Ev extends UnknownType = DefaultEventType,
  Me extends UnknownType = DefaultMessageType,
  Re extends UnknownType = DefaultReactionType,
  Us extends UnknownType = DefaultUserType
>(
  channel: Channel<At, Ch, Co, Ev, Me, Re, Us>,
) => channel.getConfig()?.commands || [];

const getMembers = <
  At extends UnknownType = DefaultAttachmentType,
  Ch extends UnknownType = DefaultChannelType,
  Co extends string = DefaultCommandType,
  Ev extends UnknownType = DefaultEventType,
  Me extends UnknownType = DefaultMessageType,
  Re extends UnknownType = DefaultReactionType,
  Us extends UnknownType = DefaultUserType
>(
  channel: Channel<At, Ch, Co, Ev, Me, Re, Us>,
) => {
  const members = (channel.state.members as unknown) as ChannelMemberResponse<
    Us
  >[];
  return members && Object.values(members).length
    ? (Object.values(members).filter((member) => member.user) as Array<
        ChannelMemberResponse<Us> & { user: UserResponse<Us> }
      >).map((member) => member.user)
    : [];
};

const getWatchers = <
  At extends UnknownType = DefaultAttachmentType,
  Ch extends UnknownType = DefaultChannelType,
  Co extends string = DefaultCommandType,
  Ev extends UnknownType = DefaultEventType,
  Me extends UnknownType = DefaultMessageType,
  Re extends UnknownType = DefaultReactionType,
  Us extends UnknownType = DefaultUserType
>(
  channel: Channel<At, Ch, Co, Ev, Me, Re, Us>,
) => {
  const watchers = (channel.state.watchers as unknown) as UserResponse<Us>[];
  return watchers && Object.values(watchers).length
    ? [...Object.values(watchers)]
    : [];
};

const getMembersAndWatchers = <
  At extends UnknownType = DefaultAttachmentType,
  Ch extends UnknownType = DefaultChannelType,
  Co extends string = DefaultCommandType,
  Ev extends UnknownType = DefaultEventType,
  Me extends UnknownType = DefaultMessageType,
  Re extends UnknownType = DefaultReactionType,
  Us extends UnknownType = DefaultUserType
>(
  channel: Channel<At, Ch, Co, Ev, Me, Re, Us>,
) => {
  const users = [...getMembers(channel), ...getWatchers(channel)];

  return Object.values(
    users.reduce((acc, cur) => {
      if (!acc[cur.id]) {
        acc[cur.id] = cur;
      }

      return acc;
    }, {} as { [key: string]: SuggestionUser<Us> }),
  );
};
type QueryMembersFunction<
  At extends UnknownType = DefaultAttachmentType,
  Ch extends UnknownType = DefaultChannelType,
  Co extends string = DefaultCommandType,
  Ev extends UnknownType = DefaultEventType,
  Me extends UnknownType = DefaultMessageType,
  Re extends UnknownType = DefaultReactionType,
  Us extends UnknownType = DefaultUserType
> = (
  channel: Channel<At, Ch, Co, Ev, Me, Re, Us>,
  query: SuggestionUser<Us>['name'],
  onReady?: (users: SuggestionUser<Us>[]) => void,
) => Promise<void>;

// TODO: test to see if this function works as it integrated a debounce function
const queryMembers = async <
  At extends UnknownType = DefaultAttachmentType,
  Ch extends UnknownType = DefaultChannelType,
  Co extends string = DefaultCommandType,
  Ev extends UnknownType = DefaultEventType,
  Me extends UnknownType = DefaultMessageType,
  Re extends UnknownType = DefaultReactionType,
  Us extends UnknownType = DefaultUserType
>(
  channel: Channel<At, Ch, Co, Ev, Me, Re, Us>,
  query: SuggestionUser<Us>['name'],
  onReady?: (users: SuggestionUser<Us>[]) => void,
): Promise<void> => {
  if (typeof query === 'string') {
    const response = (await ((channel as unknown) as Channel).queryMembers({
      name: { $autocomplete: query },
    })) as ChannelMemberAPIResponse<Us>;

    const users: SuggestionUser<Us>[] = [];
    response.members.forEach(
      (member) => isUserResponse(member.user) && users.push(member.user),
    );
    if (onReady && users) {
      onReady(users);
    }
  }
};
export const queryMembersDebounced = debounce(queryMembers, 200, {
  leading: false,
  trailing: true,
});
export const isMentionTrigger = (trigger: Trigger): trigger is '@' =>
  trigger === '@';

export type Trigger = '/' | '@';

export type TriggerSettings<
  Co extends string = DefaultCommandType,
  Us extends UnknownType = DefaultUserType
> = {
  '/': {
    component: string | React.ComponentType<Partial<CommandsItemProps<Co>>>;
    dataProvider: (
      query: SuggestionCommand<Co>['name'],
      text: string,
      onReady?: (
        data: SuggestionCommand<Co>[],
        q: SuggestionCommand<Co>['name'],
      ) => void,
    ) => SuggestionCommand<Co>[];
    output: (
      entity: SuggestionCommand<Co>,
    ) => {
      caretPosition: string;
      key: string;
      text: string;
    };
    title: string;
  };
  '@': {
    callback: (item: SuggestionUser<Us>) => void;
    component: string | React.ComponentType<Partial<MentionsItemProps<Us>>>;
    dataProvider: (
      query: SuggestionUser<Us>['name'],
      _: string,
      onReady?: (
        data: SuggestionUser<Us>[],
        q: SuggestionUser<Us>['name'],
      ) => void,
    ) => SuggestionUser<Us>[] | Promise<void> | void;
    output: (
      entity: SuggestionUser<Us>,
    ) => {
      caretPosition: string;
      key: string;
      text: string;
    };
    title: string;
  };
};

export type ACITriggerSettingsParams<
  At extends UnknownType = DefaultAttachmentType,
  Ch extends UnknownType = DefaultChannelType,
  Co extends string = DefaultCommandType,
  Ev extends UnknownType = DefaultEventType,
  Me extends UnknownType = DefaultMessageType,
  Re extends UnknownType = DefaultReactionType,
  Us extends UnknownType = DefaultUserType
> = {
  channel: Channel<At, Ch, Co, Ev, Me, Re, Us>;
  onMentionSelectItem: (item: SuggestionUser<Us>) => void;
  autocompleteSuggestionsLimit?: number;
} & Pick<TranslationContextValue, 't'>;

/**
 * ACI = AutoCompleteInput
 *
 * DataProvider accepts `onReady` function, which will execute once the data is ready.
 * Another approach would have been to simply return the data from dataProvider and let the
 * component await for it and then execute the required logic. We are going for callback instead
 * of async-await since we have debounce function in dataProvider. Which will delay the execution
 * of api call on trailing end of debounce (lets call it a1) but will return with result of
 * previous call without waiting for a1. So in this case, we want to execute onReady, when trailing
 * end of debounce executes.
 */
export const ACITriggerSettings = <
  At extends UnknownType = DefaultAttachmentType,
  Ch extends UnknownType = DefaultChannelType,
  Co extends string = DefaultCommandType,
  Ev extends UnknownType = DefaultEventType,
  Me extends UnknownType = DefaultMessageType,
  Re extends UnknownType = DefaultReactionType,
  Us extends UnknownType = DefaultUserType
>({
  channel,
  onMentionSelectItem,
  t = (msg: string) => msg,
  autocompleteSuggestionsLimit = 10,
}: ACITriggerSettingsParams<At, Ch, Co, Ev, Me, Re, Us>): TriggerSettings<
  Co,
  Us
> => ({
  '/': {
    component: 'CommandsItem',
    dataProvider: (query, text, onReady) => {
      if (text.indexOf('/') !== 0) return [];

      const selectedCommands = getCommands(channel).filter(
        (command) => query && command.name?.indexOf(query) !== -1,
      );

      // sort alphabetically unless the you're matching the first char
      selectedCommands.sort((a, b) => {
        let nameA = a.name?.toLowerCase() || '';
        let nameB = b.name?.toLowerCase() || '';
        if (query && nameA.indexOf(query) === 0) {
          nameA = `0${nameA}`;
        }
        if (query && nameB.indexOf(query) === 0) {
          nameB = `0${nameB}`;
        }
        if (nameA < nameB) return -1;
        if (nameA > nameB) return 1;

        return 0;
      });

      const result = selectedCommands.slice(0, autocompleteSuggestionsLimit);

      if (onReady) {
        onReady(result, query);
      }

      return result;
    },
    output: (entity) => ({
      caretPosition: 'next',
      key: `${entity.name}`,
      text: `/${entity.name}`,
    }),
    title: t('Commands'),
  },
  '@': {
    callback: (item) => {
      onMentionSelectItem(item);
    },
    component: 'MentionsItem',
    dataProvider: (query, _, onReady) => {
      const members = channel.state.members;

      /**
       * By default, we return maximum 100 members via queryChannels api call.
       * Thus it is safe to assume, that if number of members in channel.state is < 100,
       * then all the members are already available on client side and we don't need to
       * make any api call to queryMembers endpoint.
       */
      if (!query || Object.values(members).length < 100) {
        const users = getMembersAndWatchers(channel);

        const matchingUsers = users.filter((user) => {
          if (!query) return true;
          if (user.name?.toLowerCase().indexOf(query.toLowerCase()) !== -1) {
            return true;
          }
          if (user.id.toLowerCase().indexOf(query.toLowerCase()) !== -1) {
            return true;
          }
          return false;
        });

        const data = matchingUsers.slice(0, autocompleteSuggestionsLimit);

        if (onReady) {
          onReady(data, query);
        }

        return data;
      }
      return (queryMembersDebounced as DebouncedFunc<
        QueryMembersFunction<At, Ch, Co, Ev, Me, Re, Us>
      >)(channel, query, (data) => {
        if (onReady) {
          onReady(data, query);
        }
      });
    },
    output: (entity) => ({
      caretPosition: 'next',
      key: entity.id,
      text: `@${entity.name || entity.id}`,
    }),
    title: t('Searching for people'),
  },
});

export const MESSAGE_ACTIONS = {
  delete: 'delete',
  edit: 'edit',
  reactions: 'reactions',
  reply: 'reply',
};

export const makeImageCompatibleUrl = (url: string) =>
  (url.indexOf('//') === 0 ? `https:${url}` : url).trim();
