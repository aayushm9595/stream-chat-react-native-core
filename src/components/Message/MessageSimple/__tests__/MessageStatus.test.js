import React from 'react';
import { cleanup, render, waitFor } from '@testing-library/react-native';

import { generateMessage } from 'mock-builders/generator/message';
import { generateStaticUser, generateUser } from 'mock-builders/generator/user';
import { getTestClientWithUser } from 'mock-builders/mock';

import { MessageStatus } from '../MessageStatus';

import { Chat } from '../../../Chat/Chat';

import { Streami18n } from '../../../../utils/Streami18n';

let chatClient;
let id;
let i18nInstance;

describe('MessageStatus', () => {
  beforeAll(async () => {
    id = 'testID';
    chatClient = await getTestClientWithUser({ id });
    i18nInstance = new Streami18n();
  });
  afterEach(cleanup);

  it('should render message status with spacer', async () => {
    const user = generateUser();
    const message = generateMessage({ user });

    const { getByTestId } = render(
      <Chat client={chatClient} i18nInstance={i18nInstance}>
        <MessageStatus
          lastReceivedId={message.id}
          message={message}
          readBy={[{ id }]}
        />
      </Chat>,
    );

    await waitFor(() => {
      expect(getByTestId('spacer')).toBeTruthy();
    });
  });

  it('should render message status with delivered container', async () => {
    const user = generateUser();
    const message = generateMessage({ user });

    const { getByTestId } = render(
      <Chat client={chatClient} i18nInstance={i18nInstance}>
        <MessageStatus
          lastReceivedId={message.id}
          message={{ ...message, status: 'received' }}
          readBy={[{ id }]}
        />
      </Chat>,
    );

    await waitFor(() => {
      expect(getByTestId('delivered-container')).toBeTruthy();
    });
  });

  it('should render message status with read by container', async () => {
    const user = generateUser();
    const message = generateMessage({ user });

    const { getByTestId, getByText, rerender, toJSON } = render(
      <Chat client={chatClient} i18nInstance={i18nInstance}>
        <MessageStatus
          lastReceivedId={message.id}
          message={message}
          readBy={[{ id }, { id: user.id, name: user.name }]}
        />
      </Chat>,
    );

    await waitFor(() => {
      expect(getByTestId('avatar-text')).toBeTruthy();
      expect(getByTestId('read-by-container')).toBeTruthy();
      expect(getByText(user.name.charAt(0))).toBeTruthy();
    });

    const staticUser = generateStaticUser(0);
    const staticMessage = generateMessage({ staticUser });

    rerender(
      <Chat client={chatClient} i18nInstance={i18nInstance}>
        <MessageStatus
          lastReceivedId={staticMessage.id}
          message={staticMessage}
          readBy={[
            { id },
            {
              id: staticUser.id,
              image: staticUser.image,
              name: staticUser.name,
            },
          ]}
        />
      </Chat>,
    );

    await waitFor(() => {
      expect(toJSON()).toMatchSnapshot();
      expect(getByTestId('avatar-image')).toBeTruthy();
      expect(getByTestId('read-by-container')).toBeTruthy();
    });
  });

  it('should render message status with sending container', async () => {
    const user = generateUser();
    const message = generateMessage({ user });

    const { getByTestId } = render(
      <Chat client={chatClient} i18nInstance={i18nInstance}>
        <MessageStatus message={{ ...message, status: 'sending' }} />
      </Chat>,
    );

    await waitFor(() => {
      expect(getByTestId('sending-container')).toBeTruthy();
    });
  });
});
