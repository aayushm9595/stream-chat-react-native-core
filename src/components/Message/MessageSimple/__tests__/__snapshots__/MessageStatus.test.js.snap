// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MessageStatus should render message status with read by container 1`] = `
<View
  style={
    Array [
      Object {
        "flexDirection": "row",
        "justifyContent": "center",
        "width": 20,
      },
    ]
  }
>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "flexDirection": "row",
        },
      ]
    }
    testID="read-by-container"
  >
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
          },
        ]
      }
    >
      <Image
        accessibilityLabel="initials"
        onError={[Function]}
        resizeMethod="resize"
        size={16}
        source={
          Object {
            "uri": "https://images-na.ssl-images-amazon.com/images/I/51t29lLkg8L._AC_SL1000_.jpg",
          }
        }
        style={
          Array [
            Object {
              "borderRadius": 8,
              "height": 16,
              "width": 16,
            },
          ]
        }
        testID="avatar-image"
      />
    </View>
  </View>
</View>
`;
